### PR TITLE
Add error_description/uri info to OAuthHandler

### PR DESCRIPTION
**Pull request by HaoK**
cc @Tratcher

### CODE REVIEW

**Comment by HaoK** [comment]
Fixes https://github.com/aspnet/Security/issues/512

**Comment by PinpointTownes** [comment]
Curious: is this the final design for logging error messages? IMHO, it feels very weird compared to the great semantic logging you're adding everywhere. Why not logging an error with the (optional) error_description and error_uri and just return a minimal `AuthenticateResult`?

FWIW, we started using this pattern in the OAuth2 social providers that override `ExchangeCodeAsync` and it really seems clearer: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/76c146f3bde4d2de73e34f61ecac22cd764eb981/src/AspNet.Security.OAuth.Fitbit/FitbitAuthenticationHandler.cs#L73-L79

**Comment by HaoK** [comment]
The original motivation for these error handling changes was to actually enable the ability to do something other than just log.  Just logging should be more or less unaffected, we can certainly tweak the format of the logging, my hope was to push the logging into the base RemoteHandler which logs all errors at Info already today.

https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNet.Authentication/RemoteAuthenticationHandler.cs#L28

**Comment by HaoK** [comment]
So logging and then returning an error will result in double logging of the errors...

File: src/Microsoft.AspNet.Authentication.OAuth/OAuthHandler.cs [hunk 1]
```diff
@@ -38,7 +38,20 @@ public OAuthHandler(HttpClient backchannel)
             var error = query["error"];
             if (!StringValues.IsNullOrEmpty(error))
             {
-                return AuthenticateResult.Failed(error);
+                var errorMessage = new StringBuilder();
+                errorMessage.Append(error);
+                var errorDescription = query["error_description"];
+                if (!StringValues.IsNullOrEmpty(errorDescription))
+                {
+                    errorMessage.Append(";Description=" + errorDescription);
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.OAuth/OAuthHandler.cs]: I'm amused that you use both `Append` and `+` on the same line.

**Comment by Tratcher** [comment]
:shipit:

File: src/Microsoft.AspNet.Authentication.OAuth/OAuthHandler.cs (hunk 1 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.OAuth/OAuthHandler.cs]: Oops, switched to Append :)

**Comment by HaoK** [comment]
620622f260ba235c864d753262c0e6e923db93a5

**Pull request closed without merging**